src/PIL/JpegImagePlugin.py

```diff
@@ -497,6 +467,40 @@ def load_djpeg(self) -> None:
     def _getexif(self) -> dict[str, Any] | None:
         return _getexif(self)
 
+    def _read_dpi_from_exif(self) -> None:
+        # If DPI isn't in JPEG header, fetch from EXIF
+        if "dpi" in self.info or "exif" not in self.info:
+            return
+        try:
+            exif = self.getexif()
+            resolution_unit = exif[0x0128]
+            x_resolution = exif[0x011A]
+            try:
+                dpi = float(x_resolution[0]) / x_resolution[1]
+            except TypeError:
+                dpi = x_resolution
+            if math.isnan(dpi):
+                msg = "DPI is not a number"
+                raise ValueError(msg)
+            if resolution_unit == 3:  # cm
+                # 1 dpcm = 2.54 dpi
+                dpi *= 2.54
+            self.info["dpi"] = dpi, dpi
+        except (
+            struct.error,
+            KeyError,
+            SyntaxError,
+            TypeError,
+            ValueError,
+            ZeroDivisionError,
+        ):
+            # struct.error for truncated EXIF
+            # KeyError for dpi not included
+            # SyntaxError for invalid/unreadable EXIF
+            # ValueError or TypeError for dpi being an invalid float
+            # ZeroDivisionError for invalid dpi rational value
```

Comment:
Not sure if this is better? Feel free to merge without this.
```suggestion
            struct.error,  # truncated EXIF
            KeyError,  # dpi not included
            SyntaxError,  # invalid/unreadable EXIF
            TypeError,  # dpi is an invalid float
            ValueError,  # dpi is an invalid float
            ZeroDivisionError,  # invalid dpi rational value
        ):
```